### DISCUSSION TITLE
JSqlParser Features

### DISCUSSION THREAD

**Discussion by nitinp11** [JSQLParser/JSqlParser#1677] [Q&A]
Hi Team,
 
I am working on a task where I am reading more than 800 sqls and these sqls are having very long in nature and each sql is having keywords like case/when , Union , nested select , group by , where , with cte , joins etc.
Can this library parse bigger sql ? and after parsing can I get information like how many select statements are there and from that select how many select columns , how many tables from From clause , how many condition in AND/OR.
I need to get all such information.
I am new to this library , Can you kindly help to get this information? Is it possible and can i get this information?
 
Many Thanks

**Comment by manticore-projects** [comment]
Greetings!

> I am working on a task where I am reading more than 800 sqls and these sqls are having very long in nature and each sql is having keywords like case/when , Union , nested select , group by , where , with cte , joins etc. Can this library parse bigger sql ? 

Yes.

> and after parsing can I get information like how many select statements are there and from that select how many select columns , how many tables from From clause

Yes.

> , how many condition in AND/OR

Yes.

> . I need to get all such information. I am new to this library , Can you kindly help to get this information? Is it possible and can i get this information?

Yes.

Please refer to https://manticore-projects.com/JSQLParser/usage.html for details.

**Reply by nitinp11** [reply]
Many Thanks for your quick reply , will explore the given link and will come back for any queries.
Thanks

**Reply by nitinp11** [reply]
Is there any sample or generic code available from which i get nested (select,from,where) attribute details in one object and its parent select (select,from attribute) attribute details in parent object ,  so that once i save in DB i can get which select has which attributes. Because the examples which i am seeing are of PlainSelect etc. 
Thanks.

Sample Queries Below:

SELECT name, cost, city
FROM product 
WHERE NOT EXISTS ( SELECT id 
  FROM sale WHERE year=2020 AND product_id=product.id );

======================================================================

SELECT city, sum_price 
 FROM 
(
  SELECT city, SUM(price) AS sum_price FROM sale
  GROUP BY city
) AS s
WHERE sum_price < 2100;

=========================================================

select y.CategoryID, 
    y.CategoryName,
    round(x.actual_unit_price, 2) as "Actual Avg Unit Price",
    round(y.planned_unit_price, 2) as "Would-Like Avg Unit Price"
from
(
    select avg(a.UnitPrice) as actual_unit_price, c.CategoryID
    from order_details as a
    inner join products as b on b.ProductID = a.ProductID
    inner join categories as c on b.CategoryID = c.CategoryID
    group by c.CategoryID
) as x
inner join 
(
    select a.CategoryID, b.CategoryName, avg(a.UnitPrice) as planned_unit_price
    from products as a
    inner join categories as b on b.CategoryID = a.CategoryID
    group by a.CategoryID
) as y on x.CategoryID = y.CategoryID

**Reply by manticore-projects** [reply]
I suggest to [use this link](http://jsqlformatter.manticore-projects.com/jsqlformatter/demo.html?args=-c%20MoUQMiDCAqAEsDsCGBbApgKHt7AaWAxgPYDOALljnoQJZkCeGAYgEoDyAsrAA4BORAEwCuBCgHUAEiBYhYAOTZwQADQCSwaMFgAKeKAgxYNAZSpnzrTrBJIANpnOOqk6bPpokvWAF5YAJgAGQNMnUPgAQTkAER5+YVEAfWMfWMERMgA6ZIBKDABuDEL9KDhaBhCcfBIhFAS+GgJMSy5dWGLDMsYwqnxgGt16xths2HCtatrBhwt2Lht7CuwAcXYAVQAFWAAhAE1O4dHxjBcZaxq63ga0WAAefwBGAID8wox20voMgiQyNABzIi8ejGRbwfCfb6-AFA5DoUGwfAsIhCBACXQADwySFEQjsCRRdAuV3wfgOY1gACJwji7KMAG5-WCrBB0WDrS6NCnwxHI1G6T7cWxIBAINACfEsshExokslaCliZG2AQAWjANAA1tdwgymZK2Ry0Fzmjo9OAShEGbokBkCWQpnLYNiyLjbBLCVN4Y58AQvj9-oDgSZHCbAQI0LwEuGyEgaLYSIcnV7sKo5HJpLAAFJsVOpeJkBPkgBGybMbDksCLGT4aVEyV8Npr+ZB3VT6ZYWZzFchAcuaELWgIpao5crfqhgfrhHHvaD8JWbA22z2vp70KDjvRITbGezudaZoMpRta8nwe6F7BY9PMNQ00v3XwOr+1ttkodI3JguFovFdul94PvAJpNukA5JkBjg7h2e7dv60I0P2iYlpBYSjlWN4bg2M7ri2qGwAuS67E6OFno6XTmKOmKYVOELwWe+RAA)

It will visualize the Java Object Tree and the Hierarchy of classes ("Show AST" Button):

![image](https://user-images.githubusercontent.com/18080123/204729616-96afa3d5-df58-48e7-aca2-4539a7f71328.png)

**Reply by nitinp11** [reply]
Thanks for reply, will explore.

**Reply by nitinp11** [reply]
This is perfectly giving the AST tree , but in order to get attributes of complex queries , do i need to write my own visitor or how to use existing ? Because i am trying to understand how it is generating a good tree and once it generates in order to traverse how to know how many selects/subselects/expressions are there. 
I mean how to traverse to all the elements , is there any code sample available ? 
Please reply.

Thanks

**Reply by nitinp11** [reply]
Request you to please reply

**Reply by manticore-projects** [reply]
Greetings.

I have sent you a [link showing code examples](https://manticore-projects.com/JSQLParser/usage.html), how to traverse the AST and how to access the Objects.
I have also shown the visualisation for any AST, which will guide you.

You will need to ask much more specific questions, when you need help beyond that.
